pkg/webhooks/podimagespec/podimagespec.go

```diff
@@ -0,0 +1,293 @@
+)
+
+// PodImageSpecWebhook mutates an image spec in a pod
+type PodImageSpecWebhook struct {
+	s          *runtime.Scheme
+	kubeClient client.Client
+}
+
+// NewWebhook creates the new webhook
+func NewWebhook() *PodImageSpecWebhook {
+	scheme := runtime.NewScheme()
+	configv1.Install(scheme)
+	imagestreamv1.Install(scheme)
+	return &PodImageSpecWebhook{
+		s: scheme,
+	}
+}
+
+// Authorized implements Webhook interface
+func (s *PodImageSpecWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	ret := s.authorized(request)
+	if err := ret.Complete(request); err != nil {
+		log.Error(err, "Failed to complete the request")
+		return admissionctl.Errored(http.StatusInternalServerError, err)
+	}
+	return ret
+}
+
+func (s *PodImageSpecWebhook) authorized(request admissionctl.Request) admissionctl.Response {
+	var err error
+	ctx := context.Background()
+
+	if s.kubeClient == nil {
+		s.kubeClient, err = k8sutil.KubeClient(s.s)
+		if err != nil {
+			return admissionctl.Errored(http.StatusBadRequest, err)
+		}
+	}
+
+	pod, err := s.renderPod(request)
+	if err != nil {
+		log.Error(err, "couldn't render a Pod from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+
+	if !podContainsContainerRegexMatch(pod) {
+		return admissionctl.Allowed("Pod image spec is valid")
+	}
+
+	registryAvailable, err := s.checkImageRegistryStatus(ctx)
+	if err != nil {
+		log.Error(err, "failed to check image registry status")
+		return admissionctl.Errored(http.StatusInternalServerError, err)
+	}
+
+	if registryAvailable {
+		return admissionctl.Allowed("Image registry is available, no mutation required")
+	}
+
+	mutatedPod, err := s.mutatePod(pod, ctx)
+	if err != nil {
+		log.Error(err, "Unable mutate pod")
+		return admissionctl.Errored(http.StatusInternalServerError, err)
+	}
+
+	return admissionctl.PatchResponseFromRaw(request.Object.Raw, mutatedPod)
+}
+
+// renderPod renders the Pod in the admission Request
+func (s *PodImageSpecWebhook) renderPod(request admissionctl.Request) (*corev1.Pod, error) {
+	decoder, err := admissionctl.NewDecoder(s.s)
+	if err != nil {
+		return nil, err
+	}
+	pod := &corev1.Pod{}
+	err = decoder.Decode(request, pod)
+	if err != nil {
+		return nil, err
+	}
+	return pod, nil
+}
+
+func podContainsContainerRegexMatch(pod *corev1.Pod) (podMatch bool) {
+	podMatch = false
+
+	for i := range pod.Spec.Containers {
+		containerMatch, namespace, _, _ := checkContainerImageSpecByRegex(pod.Spec.Containers[i].Image)
+		if containerMatch && namespace == "openshift" {
+			podMatch = true
+		}
+	}
+
+	for i := range pod.Spec.InitContainers {
+		containerMatch, namespace, _, _ := checkContainerImageSpecByRegex(pod.Spec.InitContainers[i].Image)
+		if containerMatch && namespace == "openshift" {
+			podMatch = true
+		}
+	}
+
+	return
+}
+
+func (s *PodImageSpecWebhook) mutatePod(pod *corev1.Pod, ctx context.Context) ([]byte, error) {
```

Comment:
```suggestion
func (s *PodImageSpecWebhook) mutatePod(ctx context.Context, pod *corev1.Pod) ([]byte, error) {
```

nit, by convention context should be the first argument